src/ring/middleware/default_charset.clj

```diff
@@ -0,0 +1,31 @@
+(ns ring.middleware.default-charset
+  "Middleware for automatically adding a charset to the content-type header in
+  response maps."
+  (:require [ring.util.response :as response]))
+
+(defn- text-based-content-type? [content-type]
+  (or (re-find #"text/" content-type)
+      (re-find #"application/(xml|json)" content-type)))
+
+(defn- does-not-contain-charset? [content-type]
```

Comment:
Perhaps change this to:

```clojure
(defn- contains-charset? [content-type]
  (re-find #";\s*charset=[^;]*" content-type))
```

And then use `(not (contains-charset? content-type))` in the later condition.